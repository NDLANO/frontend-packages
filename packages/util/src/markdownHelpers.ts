/**
 * Copyright (c) 2021-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

//@ts-ignore
import { Remarkable } from 'remarkable';
import parse from 'html-react-parser';

export type ParseType = 'caption' | 'body';

export const parseMarkdown = (content: string, parser: ParseType = 'caption') => {
  const markdown = new Remarkable({ breaks: true, html: true });
  markdown.inline.ruler.enable(['sub', 'sup']);
  if (parser === 'caption') {
    markdown.set({ html: false, breaks: false });
    markdown.block.ruler.disable([
      'table',
      'footnote',
      'blockquote',
      'code',
      'fences',
      'footnote',
      'heading',
      'hr',
      'htmlblock',
      'lheading',
      'list',
      'table',
    ]);
    markdown.inline.ruler.disable([
      'autolink',
      'backticks',
      'del',
      'entity',
      'escape',
      'footnote_ref',
      'htmltag',
      'links',
      'newline',
      'text',
    ]);
  }
  const rendered = markdown.render(content);
  const parsed = parse(rendered);

  return parsed;
};
