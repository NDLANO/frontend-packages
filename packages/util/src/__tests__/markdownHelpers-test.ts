/**
 * Copyright (c) 2022-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/* eslint-env jest */

import { parseMarkdown } from '../markdownHelpers';

const richMarkdown =
  '### Header\n **bold** \n_italic_ \n`code` \n^sup^ \n~sub~ \nNormal list: \n* One \n* Two \n* Three \nOrdered list: \n1. 1 \n2. 2 \n3. 3';

test('parseMarkdown renders body correctly.', () => {
  const parsed = parseMarkdown(richMarkdown, 'body');

  expect(parsed).toMatchSnapshot();
});

test('parseMarkdown does not render richText when type=caption.', () => {
  const parsed = parseMarkdown(richMarkdown, 'caption');

  expect(parsed).toMatchSnapshot();
});
