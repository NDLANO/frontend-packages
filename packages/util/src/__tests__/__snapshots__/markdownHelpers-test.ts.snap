// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseMarkdown does not render richText when type=caption. 1`] = `
Array [
  <p>
    ### Header
 
    <strong>
      bold
    </strong>
     

    <em>
      italic
    </em>
     
\`code\` 

    <sup>
      sup
    </sup>
     

    <sub>
      sub
    </sub>
     
Normal list: 
* One 
* Two 
* Three 
Ordered list: 
1. 1 
2. 2 
3. 3
  </p>,
  "
",
]
`;

exports[`parseMarkdown renders body correctly. 1`] = `
Array [
  <h3>
    Header
  </h3>,
  "
",
  <p>
    <strong>
      bold
    </strong>
    <br />
    

    <em>
      italic
    </em>
    <br />
    

    <code>
      code
    </code>
    <br />
    

    <sup>
      sup
    </sup>
    <br />
    

    <sub>
      sub
    </sub>
    <br />
    
Normal list:
  </p>,
  "
",
  <ul>
    

    <li>
      One
    </li>
    

    <li>
      Two
    </li>
    

    <li>
      Three
      <br />
      
Ordered list:
    </li>
    

  </ul>,
  "
",
  <ol>
    

    <li>
      1
    </li>
    

    <li>
      2
    </li>
    

    <li>
      3
    </li>
    

  </ol>,
  "
",
]
`;
